src/vanilla.ts

```diff
@@ -43,177 +15,241 @@ type Op =
-  version: number,
-  target: T,
-  receiver: any
-): T => {
-  const cache = snapshotCache.get(receiver)
-  if (cache?.[0] === version) {
-    return cache[1] as T
-  }
-  const snapshot: any = Array.isArray(target)
-    ? []
-    : Object.create(Object.getPrototypeOf(target))
-  markToTrack(snapshot, true) // mark to track
-  snapshotCache.set(receiver, [version, snapshot])
-  Reflect.ownKeys(target).forEach((key) => {
-    const value = Reflect.get(target, key, receiver)
-    if (refSet.has(value)) {
-      markToTrack(value, false) // mark not to track
-      snapshot[key] = value
-    } else if (value instanceof Promise) {
-      if (PROMISE_RESULT in value) {
-        snapshot[key] = (value as any)[PROMISE_RESULT]
-      } else {
-        const errorOrPromise = (value as any)[PROMISE_ERROR] || value
-        Object.defineProperty(snapshot, key, {
-          get() {
-            if (PROMISE_RESULT in value) {
-              return (value as any)[PROMISE_RESULT]
-            }
-            throw errorOrPromise
-          },
-        })
-      }
-    } else if (value?.[SNAPSHOT]) {
-      snapshot[key] = value[SNAPSHOT]
-    } else {
-      snapshot[key] = value
+/**
+ * This is not a public API.
+ * It can be changed without any notice.
+ */
+export type INTERNAL_Snapshot<T> = T extends AnyFunction
+  ? T
+  : T extends AsRef
+  ? T
+  : T extends Promise<infer V>
+  ? INTERNAL_Snapshot<V>
+  : {
+      readonly [K in keyof T]: INTERNAL_Snapshot<T[K]>
     }
-  })
-  return Object.freeze(snapshot)
-}
 
-let globalVersion = 1
+// shared state
+const refSet = new WeakSet()
+const VERSION = __DEV__ ? Symbol('VERSION') : Symbol()
+const LISTENERS = __DEV__ ? Symbol('LISTENERS') : Symbol()
+const SNAPSHOT = __DEV__ ? Symbol('SNAPSHOT') : Symbol()
 
-export function proxy<T extends object>(initialObject: T = {} as T): T {
-  if (!isObject(initialObject)) {
-    throw new Error('object required')
-  }
-  const found = proxyCache.get(initialObject) as T | undefined
-  if (found) {
-    return found
-  }
-  let version = globalVersion
-  const listeners = new Set<Listener>()
-  const notifyUpdate = (op: Op, nextVersion = ++globalVersion) => {
-    if (version !== nextVersion) {
-      version = nextVersion
-      listeners.forEach((listener) => listener(op, nextVersion))
+const buildProxyFunction = (
+  objectIs = Object.is,
+
+  newProxy = <T extends object>(target: T, handler: ProxyHandler<T>): T =>
+    new Proxy(target, handler),
+
+  canProxy = (x: unknown) =>
+    isObject(x) &&
+    !refSet.has(x) &&
+    (Array.isArray(x) || !(Symbol.iterator in x)) &&
+    !(x instanceof WeakMap) &&
+    !(x instanceof WeakSet) &&
+    !(x instanceof Error) &&
+    !(x instanceof Number) &&
+    !(x instanceof Date) &&
+    !(x instanceof String) &&
+    !(x instanceof RegExp) &&
+    !(x instanceof ArrayBuffer),
+
+  PROMISE_RESULT = __DEV__ ? Symbol('PROMISE_RESULT') : Symbol(),
+  PROMISE_ERROR = __DEV__ ? Symbol('PROMISE_ERROR') : Symbol(),
```

Comment:
Should these be moved to `shared state` and be next to `VERSION`, `LISTENERS`, `SNAPSHOT`.

Reply:
Is it also possible to pass `VERSION`, `LISTENERS`, `SNAPSHOT` into `buildProxyFunction` so that a custom `createSnapshot` and custom `proxyFunction` could access these methods, but they're not directly exported?